hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/containerlog/parser/ContainerLogFileParser.java

```diff
@@ -0,0 +1,222 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Parses container log files and stores container details into a database.
+ * Uses multithreading to process multiple log files concurrently.
+ */
+
+public class ContainerLogFileParser {
+
+  private ExecutorService executorService;
+  private static final int MAX_OBJ_IN_LIST = 5000;
+
+  private static final String LOG_FILE_MARKER = ".log.";
+  private static final String LOG_LINE_SPLIT_REGEX = " \\| ";
+  private static final String KEY_VALUE_SPLIT_REGEX = "=";
+  private static final String KEY_ID = "ID";
+  private static final String KEY_BCSID = "BCSID";
+  private static final String KEY_STATE = "State";
+  private static final String KEY_INDEX = "Index";
+  private final AtomicBoolean hasErrorOccurred = new AtomicBoolean(false);
+  
+  /**
+   * Scans the specified log directory, processes each file in a separate thread.
+   * Expects each log filename to follow the format: dn-container-<roll over number>.log.<datanodeId>
+   *
+   * @param logDirectoryPath Path to the directory containing container log files.
+   * @param dbstore Database object used to persist parsed container data.
+   * @param threadCount Number of threads to use for parallel processing.
+   */
+
+  public void processLogEntries(String logDirectoryPath, ContainerDatanodeDatabase dbstore, int threadCount)
+      throws SQLException {
+    try (Stream<Path> paths = Files.walk(Paths.get(logDirectoryPath))) {
+
+      List<Path> files = paths.filter(Files::isRegularFile).collect(Collectors.toList());
+
+      executorService = Executors.newFixedThreadPool(threadCount);
+
+      CountDownLatch latch = new CountDownLatch(files.size());
+      for (Path file : files) {
+        Path fileNamePath = file.getFileName();
+        String fileName = (fileNamePath != null) ? fileNamePath.toString() : "";
+        
+        int pos = fileName.indexOf(LOG_FILE_MARKER);
+        if (pos == -1) {
+          System.out.println("Filename format is incorrect (missing .log.): " + fileName);
+          continue;
+        }
+        
+        String datanodeId = fileName.substring(pos + 5);
+        
+        if (datanodeId.isEmpty()) {
+          System.out.println("Filename format is incorrect, datanodeId is missing or empty: " + fileName);
+          continue;
+        }
+        
+        executorService.submit(() -> {
+
+          String threadName = Thread.currentThread().getName();
+          try {
+            System.out.println(threadName + " is starting to process file: " + file.toString());
+            processFile(file.toString(), dbstore, datanodeId);
+          } catch (Exception e) {
+            System.err.println("Thread " + threadName + " is stopping to process the file: " + file.toString() +
+                " due to SQLException: " + e.getMessage());
+            hasErrorOccurred.set(true);
+            executorService.shutdown();
```

Comment:
We should remove "executorService.shutdown()" from here as calling shutdown() from within a thread can stop other running tasks, this could lead other threads not completing properly.
We are already doing executorService.shutdown() after latch.await().